### ISSUE TITLE
DUCKDB_COLUMNS.DATA_TYPE_ID should reference actual STRUCT type, not generic STRUCT type

### ISSUE THREAD

**Issue by lukaseder** [duckdb/duckdb#8838]
### What happens?

In my opinion, this is a "bug", but feel free to re-classify as a change request if you think this works as expected, though I don't find the current behaviour useful.

When creating a `STRUCT` type and referencing it from a table, then the `DUCKDB_COLUMNS` view doesn't link to the actual `STRUCT` type, but to a dummy internal one.

### To Reproduce

```sql
create type u as struct (i int, j int);
create table t (u u);
```

Now do this:

```sql
select * from duckdb_types where type_oid = (select data_type_id from duckdb_columns where table_name = 't');
```

It lists this not so useful internal type information:

```
|database_name|database_oid|schema_name|schema_oid|type_oid|type_name|type_size|logical_type|type_category|internal|labels|
|-------------|------------|-----------|----------|--------|---------|---------|------------|-------------|--------|------|
|duckdb-dummy |972         |main       |974       |100     |struct   |0        |STRUCT      |COMPOSITE    |true    |      |
```

The type name is `struct`, when in fact, it should be `u`:

```sql
select * from duckdb_types where type_name = 'u';
```

Producing:

```
|database_name|database_oid|schema_name|schema_oid|type_oid|type_name|type_size|logical_type|type_category|internal|labels|
|-------------|------------|-----------|----------|--------|---------|---------|------------|-------------|--------|------|
|duckdb-dummy |972         |main       |974       |1,413   |u        |0        |STRUCT      |COMPOSITE    |false   |      |
```

### OS:

Microsoft Windows [Version 10.0.22621.2134]

### DuckDB Version:

v0.8.2-dev4025

### DuckDB Client:

JDBC

### Full Name:

Lukas Eder

### Affiliation:

Data Geekery

### Have you tried this on the latest `main` branch?

I have tested with a main build

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

**Comment by lukaseder** [comment]
I've noticed strange new behaviour in version v0.8.2-dev4025, which wasn't present in 0.8.1:

```sql
create type u as struct (i int, j int);
create table t (u u);
drop type u restrict;
```

In 0.8.1, `RESTRICT` had the expected effect:

> SQL Error: Dependency Error: Cannot drop entry "u" because there are entries that depend on it. Use DROP...CASCADE to drop all dependents.

In v0.8.2-dev4025, `RESTRICT` seems to be ignored. Given that the link between `t` and `u` isn't maintained in the `DUCKDB_COLUMNS` view, perhaps all of this is intended?

**Comment by szarnyasg** [comment]
Someone is on a roll today 🔥. Thanks for reporting these!

**Comment by lukaseder** [comment]
> Someone is on a roll today 🔥. Thanks for reporting these!

I could do much more, but it's a matter of time and priorities. I can't invest too much time in the jOOQ/DuckDB integration given these roadblocks:
https://github.com/duckdb/duckdb/issues/created_by/lukaseder

jOOQ customers expect other things to be done first. But as I've mentioned before, if the jOOQ/DuckDB integration was sponsored, I could prioritise it a bit more :)

**Comment by jdnvn** [comment]
looking into this, but when I try to reproduce:

```
D create type e as enum ('a', 'b');
D create type u as struct (i int, j int);
D create table t (e e, u u);
D select data_type, data_type_id
> from duckdb_columns
> where table_name = 't';

┌──────────────────────────────┬──────────────┐
│          data_type           │ data_type_id │
│           varchar            │    int64     │
├──────────────────────────────┼──────────────┤
│ ENUM('a', 'b')               │          104 │
│ STRUCT(i INTEGER, j INTEGER) │          100 │
└──────────────────────────────┴──────────────┘
```

I get this. Enum has the same format for the `data_type` column. Are we sure w want `e` and `u` instead?

**Comment by lukaseder** [comment]
> looking into this, but when I try to reproduce:

Maybe this has changed already due to my recent reports in this area? I've made a few :)

> Are we sure w want `e` and `u` instead?

The main point I'm trying to make here is that the column reference to the type should reference the *user type*, not the *system type*. Maybe both are fine, too, e.g. SQL Server `sys.columns` has both `system_type_id` and `user_type_id`: https://learn.microsoft.com/en-us/sql/relational-databases/system-catalog-views/sys-columns-transact-sql?view=sql-server-ver16. But the more important reference is the *user type*

Personally, I believe that `data_type` should always reference the type name, plus additional qualification of database and schema would be necessary, too. This is how the SQL standard `INFORMATION_SCHEMA` does it. Everything is always qualified in 3 parts, e.g. for UDTs:

- `UDT_CATALOG` (e.g. database)
- `UDT_SCHEMA`
- `UDT_NAME` 

The `INFORMATION_SCHEMA` does this *everywhere*, consistently, for all identifiers!

Mixing identifiers with structural descriptions in the same column feels weird to me. As a user of such a schema, I'm always going to be worrying if I'm using it correctly, and if it will break in the future. E.g. MySQL did this for years with their messy interpretation of the `INFORMATION_SCHEMA`. They have the `COLUMNS.DATA_TYPE` and `COLUMNS.COLUMN_TYPE` columns, and without reverse engineering things on multiple versions, it's impossible to trust their contents: https://dev.mysql.com/doc/refman/8.0/en/information-schema-columns-table.html

I understand that DuckDB supports fully structural types, i.e. types that don't have a name. But those types could still have a surrogate key in the `duckdb_types` view, and hopefully also a structural representation in a new view as I've suggested here:

- https://github.com/duckdb/duckdb/discussions/8832

The more regular these things are, the more trustworthy they become, and the more they will be adopted by tooling (e.g. Dbeaver, jOOQ, etc.)
